libdnf/module/module_db.cpp

```diff
@@ -0,0 +1,346 @@
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "module/module_db.hpp"
+
+#include "base/base_impl.hpp"
+#include "utils/bgettext/bgettext-mark-domain.h"
+
+#include "libdnf/module/module_errors.hpp"
+#include "libdnf/module/module_query.hpp"
+
+#include <modulemd-2.0/modulemd.h>
+
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <set>
+#include <string>
+
+namespace libdnf::module {
+
+
+bool ModuleDB::add_runtime_module_state(const std::string & module_name) {
+    // There can only be one config per module
+    if (runtime_module_states.find(module_name) != runtime_module_states.end()) {
+        return false;
+    }
+
+    auto runtime_module_state = RuntimeModuleState();
+    runtime_module_state.status = ModuleStatus::AVAILABLE;
+    runtime_module_state.stream_changes_num = 0;
+
+    try {
+        // First, try to get the modules state from config.
+        system::ModuleState config_module_state = base->p_impl->get_system_state().get_module_state(module_name);
+        runtime_module_state.status = config_module_state.status;
+        if (runtime_module_state.status == ModuleStatus::ENABLED) {
+            runtime_module_state.enabled_stream = config_module_state.enabled_stream;
+            runtime_module_state.installed_profiles = config_module_state.installed_profiles;
+        }
+    } catch (system::StateNotFoundError &) {
+        // If config doesn't exist, check if the module even exists.
+        ModuleQuery query(base, false);
+        query.filter_name(module_name);
+        if (query.empty()) {
+            throw NoModuleError(M_("No such module: {}"), module_name);
+        }
+    }
+
+    runtime_module_states.emplace(module_name, runtime_module_state);
+    return true;
+}
+
+
+const ModuleDB::RuntimeModuleState & ModuleDB::get_runtime_module_state(const std::string & module_name) const {
+    try {
+        return runtime_module_states.at(module_name);
+    } catch (std::out_of_range &) {
+        throw NoModuleError(M_("No such module: {}"), module_name);
+    }
+}
+
+
+ModuleDB::RuntimeModuleState & ModuleDB::get_runtime_module_state(const std::string & module_name) {
+    try {
+        return runtime_module_states.at(module_name);
+    } catch (std::out_of_range &) {
+    }
+
+    add_runtime_module_state(module_name);
```

Comment:
The methods `get_runtime_module_state` calls 2x `runtime_module_states.at(module_name)` and 1X `runtime_module_states.find(module_name)` . It casts not much but it is not optimal.

Reply:
Hm, right now, there are 3 calls the first time the state is accessed, 1 call afterwards. I could either change it to 2 calls always, or change the `add_runtime_module_state` to return the object it just created or that existed before (then the `add_runtime_module_state` could be called directly, but then it probably should be renamed to not be confusing). What do you think?